### DISCUSSION TITLE
Hi

### DISCUSSION THREAD

**Discussion by terminator356** [muse-sequencer/muse#1224] [Team Posts]
Just want to say I'm still here. I've not abandoned.

**Comment by spamatica** [comment]
Hey man! Still here too :) became a bit of a hiatus... Haven't seen kybos around either, we will just have to see if he resurfaces.

How is it going with the last fixes to make a fix release? I can't remember anymore exactly what you were working on. The part extend bits I guess, was it something else too? Should I go ahead and make a release with what we have? I have to go through it and see exactly what that is. I think I noticed some new bugs that need addressing, but winter is coming and with that a bit more MusEing!

**Comment by terminator356** [comment]
Believe it or not all this time I've been working on that border resizing stuff.
And it's coming along great. Sorry for the loooong dely.

What I have done is incorporate your original idea - Auto Expansion of waves in parts.
It's now a button on the Arranger toolbar.
When active, it auto-sizes the waves so that there's no 'hidden events arrows' like there is now.
So users have a choice. Both modes are very handy, and essential really, to have.
"But wait, there's more!" the man said. You knew that was coming, eh?

As you requested: Changed part 'split' routines to leave no 'hidden events' arrows (at least on waves).

I've figured out a way to redraw the wave and midi parts as they are being resized,
 even in 'drag events with the border' mode (holding the Ctrl key while resizing a border).
It does not currently do that. Further, currently the actual transaction from the
 graphical 'item' to the underlying model (the undo operation) only happens when
 the mouse button is released - whatever Ctrl state was when released is used
 to adjust the object. So I am supporting accumulation of the Ctrl key actions.
This means a user can, while pressing the mouse button and dragging a border,
 intermittently hold Ctrl to say "drag the events with the border for a while,
 the no drag, then drag..." This is very important for the Ctrl action to accumulate
 as you go along. I use some temporary values inside each canvas item, to hold
 these dynamically changing values.

Also, bugs: I fixed several places that were calling Event::clone() when they
 should have been calling Event::duplicate(). This caused **crashes** and broke our 
 "no events with duplicate ID numbers in the same event list" rule.
(clone() copies the ID, duplicate creates a new ID).

Observed a crash, **we must fix this**:
Tempo Master Track Editor: Do not allow tempo graph to go below 0.
Actually never even below 20.
It is currently possible to draw the tempo graph below zero. Crash!

Wave Editor: Added long-missing 'Delete events' menu item.
It also sorely needs a 'Load Sample' menu command, eh?

C++17 is nice.
This change eliminates some (de)allocations in the realtime thread and boosts speed:
Changed several realtime container 'change' undo operations from erase()/insert()
 to C++17 extract()/insert(). (For example Undo::ModifyEvent).
Added new operation 'ModifyEventProperties' which changes an event's position,
 length, and wave start position all at once, using realtime-safe C++17 extract/insert.

All for now. Cheers.
